### ISSUE TITLE
Getting requested name when using wildcards

### ISSUE THREAD

**Issue by AydinHassan** [PHP-DI/PHP-DI#269]
Hi, I tried to look through the docs and code about wilcard definitions but couldn't find what I wanted. Is it possible to get the requested service name in a wildcard factory?

```php
return [
    'Vendor\Package\Command\*Command' => factory(function (Container $c, $requestedName) {
        $cmd = explode("\\", $requestedName)[3]; //DeployCommand
        $cmd = substr($cmd, 0, -7); //Deploy
        $commandService = $c->get(sprintf('%s/Service', $cmd));
        return new $requestedName($commandService);
    }),
];
```

Maybe there is a better way to do what I am trying to accomplish?

**Comment by mnapoli** [comment]
Hi! There is indeed no way to do that for now unfortunately. Let's keep that issue open, I'm currently focused on releasing 5.0 (tomorrow) ;)

**Comment by AydinHassan** [comment]
Cool :) I just had a mess about and got it working by changing

`return call_user_func($callable, $this->container);` to `return call_user_func_array($callable, [$this->container, $definition->getName()]);`

in

`\DI\Definition\Resolver\FactoryDefinitionResolver::resolve`

If you wouldn't mind the change I could PR with some tests?

**Comment by mnapoli** [comment]
I think this change is fine for factories used for wildcards definitions, but introducing this might make it incompatible with #197 (when it's implemented). I think ideally the definition name should only be passed for wildcard definitions and keep other factories unchanged (only the container as a parameter).

**Comment by AydinHassan** [comment]
hmm, okay, that makes sense. It doesn't look like there is an easy way to know if the definition was a wildcard one at that point. I will leave this open and you can point me in the direction when you're a bit less busy!

**Comment by mnapoli** [comment]
Will do! An idea I just thought of: add a list of "extra parameters" in [FactoryDefinition](https://github.com/PHP-DI/PHP-DI/blob/master/src/DI/Definition/FactoryDefinition.php):

- by default the array is empty
- for wildcard definitions the [DefinitionArray](https://github.com/PHP-DI/PHP-DI/blob/master/src/DI/Definition/Source/DefinitionArray.php) will populate the array with the real class name

When the factory is called, do something like:

```php
return call_user_func_array($callable, array_merge([$this->container], $definition->getExtraParameters());
```

What do you think?

**Comment by AydinHassan** [comment]
Yeah that works for me! That would be a BC break right?

**Comment by mnapoli** [comment]
I don't think so, where do you see it?

**Comment by AydinHassan** [comment]
Because `\DI\Definition\Resolver\FactoryDefinitionResolver::resolve` is only type hinted as receiving `\DI\Definition\Definition`. We can't expect that it is in an instance of `DI\Definition\FactoryDefinition` can we? In order to call `getExtraParameters()` we have to add it to the `\DI\Definition\Definition` interface, no?

**Comment by mnapoli** [comment]
Oh right yes! :) My idea was more of a quick hack but if it affects the interfaces and BC I don't like it so much anymore.

Maybe it would be good to implement it with #197 in a clean way

**Comment by mnapoli** [comment]
This has been implemented in #345 and #347 and will be coming in PHP-DI 5.2!

**Comment by AydinHassan** [comment]
Awesome! Nice work. I'll give this a whirl at some point :smile:
